### ISSUE TITLE
Linear evaluation weight problems ("_IncompatibleKeys" and access issues)

### ISSUE THREAD

**Issue by KnockerPulsar** [facebookresearch/dino#117]
Hello, thanks for all the amazing work you put into this.
I tried downloading the pretrained weights for the 3 available ViT models and have encountered some issues:
1. ViT-S/16 gives me an "Access denied" message whenever I try downloading it.
2. For ViT-S/8 or ViT-B/16, their weights seem corrupted? Whenever I try loading them into `eval_linear.py`, I get a message listing a long list of missing and unexpected keys. Not exactly sure what is wrong here. The loss also starts at a quite high value and although it's dropping off, I don't think this is the intended behavior from a pretrained model. 

Here's the training output (not the log) for ViT-S/8
[TrainingOutput.txt](https://github.com/facebookresearch/dino/files/7099235/TrainingOutput.txt)

Thank you again for your work.

**Comment by woctezuma** [comment]
This kind of issue (`_IncompatibleKeys`) happened in the past: #65

Maybe code was updated without updating the models, or vice-versa.

**Comment by KnockerPulsar** [comment]
I'd be willing to attempt to fix it if there's a way to see the checkpoint data. I tried opening the files in a text editor and most files seem in binary. Any clue how I could see what the keys are? I suspect something similar to 94175993abde84179449d79e22eab7ea28dec14b ...

**Comment by mathildecaron31** [comment]
@KnockerPulsar 
1) Thanks for pointing this out. There was indeed a typo in my path for downloading the linear weights obtained from evaluating the ViT-S/16 model. That should be fixed now.

2) I am not sure to understand what you seek to do exactly:

a) You can run a linear evaluation given some pre-trained backbone weights. In this case running the following command will train a fully connected linear layer on top of the backbone. You can check the logs at https://dl.fbaipublicfiles.com/dino/dino_deitsmall16_pretrain/dino_deitsmall16_pretrain_eval_linear_log.txt
```
python -m torch.distributed.launch --nproc_per_node=8 eval_linear.py --data_path /path/to/imagenet
```

b) You can directly load the linear weights and check the performance of the model if you don't want to re-train this layer.
The linear weights are provided here https://github.com/facebookresearch/dino#evaluation-linear-classification-on-imagenet while the backbone weights are provided here https://github.com/facebookresearch/dino#pretrained-models.

Anyway you can check all the checkpoint data by loading the checkpoints with pytorch:
```
import torch
checkpoint = torch.load(path_to_checkpoint)
```
Hope that helps.

**Comment by mathildecaron31** [comment]
@KnockerPulsar 

I have tried to improve the code so you can easily load the pretrained linear weights and check their performance (situation b. in my post above). For example you can run:
```
python eval_linear.py --evaluate --arch vit_small --patch_size 16 --data_path /path/to/imagenet/train
```
```
python eval_linear.py --evaluate --arch vit_small --patch_size 8 --data_path /path/to/imagenet/train
```
```
python eval_linear.py --evaluate --arch vit_base --patch_size 16 --n_last_blocks 1 --avgpool_patchtokens true --data_path /path/to/imagenet/train
```
```
python eval_linear.py --evaluate --arch vit_base --patch_size 8 --n_last_blocks 1 --avgpool_patchtokens true --data_path /path/to/imagenet/train
```
```
python eval_linear.py --evaluate --arch resnet50 --data_path /path/to/imagenet/train
```

https://github.com/facebookresearch/dino/commit/d2f3156bb34c32ae54312de7fe4e31580f12ff7f

Hope that helps.

**Comment by kikacaty** [comment]
> @KnockerPulsar
> 
> I have tried to improve the code so you can easily load the pretrained linear weights and check their performance (situation b. in my post above). For example you can run:
> 
> ```
> python eval_linear.py --evaluate --arch vit_small --patch_size 16 --data_path /path/to/imagenet/train
> ```
> 
> ```
> python eval_linear.py --evaluate --arch vit_small --patch_size 8 --data_path /path/to/imagenet/train
> ```
> 
> ```
> python eval_linear.py --evaluate --arch vit_base --patch_size 16 --n_last_blocks 1 --avgpool_patchtokens true --data_path /path/to/imagenet/train
> ```
> 
> ```
> python eval_linear.py --evaluate --arch vit_base --patch_size 8 --n_last_blocks 1 --avgpool_patchtokens true --data_path /path/to/imagenet/train
> ```
> 
> ```
> python eval_linear.py --evaluate --arch resnet50 --data_path /path/to/imagenet/train
> ```
> 
> [d2f3156](https://github.com/facebookresearch/dino/commit/d2f3156bb34c32ae54312de7fe4e31580f12ff7f)
> 
> Hope that helps.

Hi,
I tried to evaluate the lincls with vit_base and vit_small. On vit_small it works perfect but on vit_base it gave nearly 0 ACC. Is there an issue with the provided weights?

Thanks,

**Comment by mathildecaron31** [comment]
hi @kikacaty can you do a fresh clone and run again ? The problem with vit_base is due to a discrepency in the way I concatenated the features (related to https://github.com/facebookresearch/dino/issues/134) but that should be fixed in this commit https://github.com/facebookresearch/dino/commit/d2f3156bb34c32ae54312de7fe4e31580f12ff7f

![image](https://user-images.githubusercontent.com/20599397/135046936-93cde546-8f6f-49aa-8073-ab92fd36556e.png)
